File: android/app/src/main/kotlin/com/stockbook/app/feature/book/PartyList.kt
package com.stockbook.app.feature.book

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.stockbook.app.design.GhostButton
import com.stockbook.app.design.Glyph
import com.stockbook.app.design.Icon
import com.stockbook.app.design.Kicker
import com.stockbook.app.design.Metrics
import com.stockbook.app.design.Nocturne
import com.stockbook.app.design.NocturneField
import com.stockbook.app.design.NocturneType
import com.stockbook.app.design.card
import com.stockbook.core.model.Currency
import com.stockbook.core.money.Money
import com.stockbook.core.text.Strings

/**
 * One person on the list, flattened out of `Customer` or `Supplier`.
 *
 * The two are the same shape pointed in opposite directions, but they are
 * separate types in the domain and giving them a shared interface there would be
 * inventing a concept the arithmetic does not have. Flattening at the edge, where
 * the only question is what to draw, costs one data class and keeps the domain
 * honest.
 */
data class PartyRow(val key: String, val name: String, val contact: String?, val owed: Double)

/**
 * The directory at the top of each half of the Book: who the shop deals with, and
 * the way into any one of them.
 *
 * This replaced a dropdown. A customer has a balance, a statement, payments and
 * credit notes against them, and until now the only way to reach any of it was to
 * pick a name out of a filter on a list of bills — which meant that people, the
 * thing half this app is about, were the one entity with no screen. The evidence
 * that this was wrong is that Today needed a banner *and* a purpose-built sheet
 * to get from "Ahmed still owes" to Ahmed.
 *
 * Capped rather than complete. A shop with two hundred customers should not have
 * to scroll past all of them to reach today's bills, so the list shows the few
 * who owe most — [com.stockbook.core.store.StockbookStore.customers] hands them
 * over in that order — and the rest are behind the search box or the toggle.
 */
@Composable
fun PartyList(
    title: String,
    rows: List<PartyRow>,
    /** Everybody, by name, matching what has been typed. Only read while searching. */
    search: (String) -> List<PartyRow>,
    addTitle: String,
    emptyMessage: String,
    currency: Currency,
    strings: Strings,
    onAdd: () -> Unit,
    onOpen: (String) -> Unit,
    modifier: Modifier = Modifier
) {
    var query by remember { mutableStateOf("") }
    var expanded by remember { mutableStateOf(false) }

    val searching = query.isNotBlank()
    val shown = when {
        searching -> search(query)
        expanded -> rows
        else -> rows.take(VISIBLE)
    }

    Column(modifier = modifier.fillMaxWidth()) {
        Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
            Kicker(title, modifier = Modifier.weight(1f))
            GhostButton(addTitle, onClick = onAdd, fontSize = 12.0)
        }
        Spacer(Modifier.height(8.dp))

        if (rows.isEmpty()) {
            Text(emptyMessage, style = NocturneType.meta, color = Nocturne.neutral500)
            return@Column
        }

        // Offered only once the list is longer than it is worth reading through.
        // A shop with four customers does not need a way to search four names.
        if (rows.size > VISIBLE) {
            NocturneField(
                value = query,
                onValueChange = { query = it },
                placeholder = strings.search,
                height = 40.dp,
                fontSize = 13.5,
                modifier = Modifier.padding(bottom = Metrics.rowGap)
            )
        }

        if (shown.isEmpty()) {
            Text(strings.nobodyMatches, style = NocturneType.meta, color = Nocturne.neutral500)
            return@Column
        }

        for (row in shown) {
            PartyRowView(
                row = row,
                currency = currency,
                strings = strings,
                onClick = { onOpen(row.key) },
                modifier = Modifier.padding(bottom = Metrics.rowGap)
            )
        }

        // Nothing to expand while searching: the search already showed everybody
        // who answers to what was typed.
        if (!searching && !expanded && rows.size > VISIBLE) {
            GhostButton(strings.all, onClick = { expanded = true }, fontSize = 12.0)
        }
    }
}

/** How many fit above the documents before the list stops being a summary. */
private const val VISIBLE = 5

/**
 * One name and what stands against it.
 *
 * The balance is the row's whole reason to be read, so it is what the eye lands
 * on: accent where money is outstanding, neutral where it is not. "Settled up"
 * rather than a blank — the absence of a figure reads as a row that failed to
 * load.
 */
@Composable
private fun PartyRowView(
    row: PartyRow,
    currency: Currency,
    strings: Strings,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = modifier
            .fillMaxWidth()
            .card(Metrics.rowRadius)
            .clickable(onClick = onClick)
            .padding(horizontal = 12.dp, vertical = 11.dp)
    ) {
        Glyph(Icon.customer, size = 14.dp, tint = Nocturne.neutral500)
        Spacer(Modifier.width(9.dp))
        Column(modifier = Modifier.weight(1f)) {
            Text(
                row.name,
                style = NocturneType.rowPrimary,
                color = Nocturne.text,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
            row.contact?.let {
                Text(
                    it,
                    style = NocturneType.meta,
                    color = Nocturne.neutral500,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )
            }
        }
        Spacer(Modifier.width(10.dp))
        Text(
            when {
                row.owed > 0 -> Money.text(row.owed, currency)
                row.owed < 0 -> strings.inAdvance(Money.text(-row.owed, currency))
                else -> strings.settledUp
            },
            style = NocturneType.inter(13.0),
            color = if (row.owed > 0) Nocturne.accent400 else Nocturne.neutral500,
            maxLines = 1
        )
        Spacer(Modifier.width(8.dp))
        Glyph(Icon.openRow, size = 12.dp, tint = Nocturne.neutral500)
    }
}

File: ios/Stockbook/Features/Book/PartyList.swift
import SwiftUI

/// One person on the list, flattened out of `Customer` or `Supplier`.
///
/// The two are the same shape pointed in opposite directions, but they are
/// separate types in the domain and giving them a shared protocol there would be
/// inventing a concept the arithmetic does not have. Flattening at the edge,
/// where the only question is what to draw, costs one struct and keeps the
/// domain honest. The Kotlin twin does exactly this.
struct PartyRow: Identifiable {
    let key: String
    let name: String
    let contact: String?
    let owed: Double

    var id: String { key }
}

/// The directory at the top of each half of the Book: who the shop deals with,
/// and the way into any one of them.
///
/// This replaced a dropdown. A customer has a balance, a statement, payments and
/// credit notes against them, and until now the only way to reach any of it was
/// to pick a name out of a filter on a list of bills — which meant that people,
/// the thing half this app is about, were the one entity with no screen. The
/// evidence that this was wrong is that Today needed a banner *and* a
/// purpose-built sheet to get from "Ahmed still owes" to Ahmed.
///
/// Capped rather than complete. A shop with two hundred customers should not
/// **Everybody, not the top five.** The list used to stop at five names with an
/// "All" underneath, because every row was built whether or not it was on screen
/// — two hundred customers meant two hundred rows composed to show five. A
/// `LazyVStack` builds only what is visible, which removes the reason for the cap
/// rather than moving it: no page size, no button, just the list.
///
/// Nothing to fetch, either. The roster is already in memory — `customers()` is a
/// walk over a snapshot — so there is no page to load, only rows to draw, and
/// drawing them on demand is the whole of it.
///
/// Biggest debt first, which is the order `customers()` hands them over in.
struct PartyList: View {
    let title: String
    let rows: [PartyRow]
    /// Everybody, by name, matching what has been typed. Read only while searching.
    let search: (String) -> [PartyRow]
    let addTitle: String
    let emptyMessage: String
    let onAdd: () -> Void
    let onOpen: (String) -> Void

    @State private var query = ""

    /// Above how many names a search box earns its place.
    ///
    /// Not a page size — the list shows everybody. This is only the point at
    /// which scrolling stops being the quickest way to find one.
    private static let searchable = 5

    private var searching: Bool { !query.isBlank }

    private var shown: [PartyRow] { searching ? search(query) : rows }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack(spacing: 8) {
                Kicker(title)
                Spacer(minLength: 6)
                Button(addTitle) { onAdd() }
                    .buttonStyle(GhostButtonStyle(fontSize: 12))
            }
            .padding(.bottom, 8)

            if rows.isEmpty {
                Text(emptyMessage).nocturneText(.meta)
            } else {
                // Offered only once the list is longer than it is worth reading
                // through. A shop with four customers does not need a way to
                // search four names.
                if rows.count > Self.searchable {
                    NocturneField(
                        placeholder: Loc.search,
                        text: $query,
                        height: 40,
                        fontSize: 13.5
                    )
                    .padding(.bottom, Metrics.rowGap)
                }

                if shown.isEmpty {
                    Text(Loc.nobodyMatches).nocturneText(.meta)
                } else {
                    // Lazy: the enclosing `ScrollView` builds a row when it comes
                    // into view and not before, which is what lets this show the
                    // whole roster.
                    LazyVStack(spacing: Metrics.rowGap) {
                        ForEach(shown) { row in
                            Button {
                                onOpen(row.key)
                            } label: {
                                PartyRowView(row: row)
                            }
                            .buttonStyle(.plain)
                        }
                    }
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }
}

/// One name and what stands against it.
///
/// The balance is the row's whole reason to be read, so it is what the eye lands
/// on: accent where money is outstanding, neutral where it is not. "Settled up"
/// rather than a blank — the absence of a figure reads as a row that failed to
/// load.
private struct PartyRowView: View {
    let row: PartyRow

    @Environment(\.currency) private var currency

    var body: some View {
        HStack(spacing: 9) {
            Glyph(Icon.customer, size: 14)
                .foregroundStyle(Nocturne.neutral500)
            VStack(alignment: .leading, spacing: 1) {
                Text(row.name).nocturneText(.rowPrimary).lineLimit(1)
                if let contact = row.contact {
                    Text(contact).nocturneText(.meta).lineLimit(1)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            Text(balanceText)
                .font(NocturneType.inter(13))
                .foregroundStyle(row.owed > 0 ? Nocturne.accent400 : Nocturne.neutral500)
                .lineLimit(1)

            Glyph(Icon.openRow, size: 12)
                .foregroundStyle(Nocturne.neutral500)
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 11)
        .frame(maxWidth: .infinity)
        .background(Nocturne.surface, in: RoundedRectangle(cornerRadius: Metrics.rowRadius, style: .continuous))
        .contentShape(Rectangle())
    }

    private var balanceText: String {
        if row.owed > 0 { return Money.text(row.owed, in: currency) }
        if row.owed < 0 { return Loc.inAdvance(Money.text(-row.owed, in: currency)) }
        return Loc.settledUp
    }
}

extension Customer {
    /// As the directory draws it.
    var directoryRow: PartyRow {
        PartyRow(key: key, name: name, contact: contactLine, owed: owed)
    }

    /// `0500 111 222 · Al Khobar`, or nothing for somebody who is only a name.
    var contactLine: String? {
        let details = [phone, place].compactMap { $0 }
        return details.isEmpty ? nil : details.joined(separator: " · ")
    }
}

extension Supplier {
    var directoryRow: PartyRow {
        PartyRow(key: key, name: name, contact: contactLine, owed: owed)
    }

    var contactLine: String? {
        let details = [phone, place].compactMap { $0 }
        return details.isEmpty ? nil : details.joined(separator: " · ")
    }
}
